Revise bandwidth reduction page with cellular measurements

kb2ma opened the pull request:
Emphasize bandwidth used per feature, and simplify. Also add the effect of NetworkManager connectivity check.

kb2ma commented:
@vipulgupta2048 , I'm happy for any comments but most interested in the overall structure of the revision. I don't want to start tweaking before we're happy with the approach. Overall I tried to focus on actual bandwidth usage because I think that's most relevant.

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
 
-Side Effects / Warnings
--------------------
+| Control | Monthly Bandwidth if enabled | Variable Name<br>`BALENA_SUPERVISOR...` | Comments |
```
```suggestion
| Control | Monthly Bandwidth if enabled | Variable Name| Comments |
```

Much similar to write the complete variable name @kb2ma because someone would be copying these straightup and not liking the experience.

vipulgupta2048 replied:
Suggestion: replace `Comments` with `Use` or `description`

kb2ma replied:
I was trying to conserve horizontal space, but I like the practicality of your suggestion to include the full name. I'm going to try putting the variable name below the name of the feature in the leftmost cell.

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
--------------------
+| Control | Monthly Bandwidth if enabled | Variable Name<br>`BALENA_SUPERVISOR...` | Comments |
+| ------- | --------- | -------- | -------- |
+| Cloudlink VPN | 81 MB | ...`_VPN_CONTROL` | Bandwidth just to maintain connection |
```
Maybe add a column about essential or not? This seems like an absolute necessity to have? or structure the comment like this?

```suggestion
| Cloudlink VPN | 81 MB | ...`_VPN_CONTROL` | Maintains remote connection with the device for access |
```

This explains what it does, signifies importance, and justifies memory usage. If the user wants to disable any of these they should know the consequences as well as the effects.

vipulgupta2048 replied:
The description I gave is an example, feel free add a better more accurate one.

vipulgupta2048 replied:
Ahh there you go, there is a side effects section. Maybe add a oneliner above the table for the user to check the side effects as well

kb2ma replied:
Hopefully with more space, I can add more text in the Comments/Description cell. I agree it would help for the text to be a little higher level since each row is the first mention of the feature. At the same time I need to qualify what exactly was measured to set expectations.

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
 # Reduce bandwidth usage
 
-In order to provide users with essential information about their devices, the {{ $names.company.lower }} Supervisor on the device regularly keeps the API informed about the device's state. Upon receiving a target device state which differs from the current device state, the Supervisor applies changes, such as downloading new application updates or setting environment variables. While a quick resp ...
+balenaOS and the device Supervisor provide convenient features that make it easy to manage a device remotely, like forwarding container logs, hardware metrics reporting, and Cloudlink for on-demand external access. However, these features consume network bandwidth, and typically cellular connectivity is charged by the megabyte. To allow you to trade off between cost and features, you may configur ...
```
I know you didn't want any comments on content, but shouldn't this have engine too? Or engine takes up no bandwidth?

kb2ma replied:
I struggled some to be comprehensive with these categories. The engine isn't involved here. It just runs what the Supervisor has provided for it.

kb2ma replied:
Confusion on my part; nothing further to do here.

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
+Each actual variable name is prefixed with "`BALENA_SUPERVISOR`", for example `BALENA_SUPERVISOR_VPN_CONTROL`.
 
-`BALENA_SUPERVISOR_CONNECTIVITY_CHECK`: Defines the device's ability to test and indicate (via an LED when available) that it has issues with connectivity.
+You can perform these measurements yourself as well! See the Network Metrics Logger [example](https://github.com/balena-io-examples/network-metrics-logger).
```
Meta: Is it a good look for us to own the measurement tool to check our own metrics? It should be fine, if we can add a small paragraph on the "why we built this"? + If you can add alternatives that people can use to test but not as good as our checker for transparency. Showing the thinking process & what we used to build this will help the customers in the future who will be using this network logger to check their devices and find the results acceptable. 

Example: I love when my ISP can provide speed results using Ookla but I trust fast.com, ismyinternetgoodenough.com or at least some other tool to co-relate the performance.

Meta thought, feel free to skip

kb2ma replied:
Ultimately I think most/all tools report the statistics the networking subsystem provides, probably via `/proc`. Great idea to provide rationale. We build this to have a tool independent of the cellular provider. For example, Twilio does publish byte usage per device per hour, but you've got use its API. I did verify that our numbers match theirs really closely.

vipulgupta2048 replied:
SGTM, i think a rationale section be great to have

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
+| API Poll Interval, 1x day | 1 MB | ...`_POLL_INTERVAL` | Minimum interval |
+| Device Logging| ?? | ...`_LOG_CONTROL` | Minimal application activity |
+| Cloudlink connectivity check | N/A | ...`_CONNECTIVITY_CHECK`  | Only active when VPN is down |
 
```
Future visions: How are we planning to keep the data updated? Do we have a test environment spun up somewhere that can keep this table updated? Reach out to us in autokit if you are looking for an automated solution to keep this table updated with OS releases. Could be helpful?

kb2ma replied:
I do not have a plan on how to keep the data updated. The OS team has talked about using the tool periodically to verify no regressions. This capability was another of the rationales behind the work.

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
+**Device Logging** Any logs written by the user container or the device Agent are not sent to the dashboard when this variable is set to false.
+
+**Cloudlink connectivity check** Defines the device's ability to test and indicate (via an LED when available) that it has issues with connectivity.
 
```
In the end, I might like to see a section on how to apply these using the dashboard and SV API. You can either link to docs for both or write a sample request. Your choice. 

My recommendation would be the former and it will be a clear CTA

kb2ma replied:
Cool; again I appreciate the practical approach.

vipulgupta2048 commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,41 @@ title: Reduce bandwidth usage
-* Change `BALENA_SUPERVISOR_POLL_INTERVAL` to 24 hours (86400000 ms)
-* Disable `BALENA_SUPERVISOR_LOG_CONTROL`
-* Disable `BALENA_SUPERVISOR_HARDWARE_METRICS`
 
```
Could we also add a section to recommend users to use our Metrics block to check bandwidth usage if they want. Could be a good pitch to use it on their fleet or at least some devices?

Shows that we trust our results and look forward to more data to improve this? What say

kb2ma replied:
I'm a little confused here. I did suggest above that users verify for themselves. The linked project uses the Metrics block from Hub.

> You can perform these measurements yourself as well!

vipulgupta2048 replied:
That is true @kb2ma Yeah, let's leave this one out. Since there is a getting started section on the network tool

kb2ma commented:
@vipulgupta2048 , tightened up the presentation format. I responded and resolved some comments. The others remain to implement. Are you happy with the overall flow now?

vipulgupta2048 commented:
I am satisfied with ~my care~ the overall flow @kb2ma Go for it.

kb2ma commented:
Addressed all comments and refined text. Ready for review, @vipulgupta2048 .

kb2ma commented:
@rosswesleyporter , I'd appreciate your review here. See [this file](https://github.com/balena-io/docs/blob/16626b069bbbc226d4ec61d6c0a2241880b0c9a6/pages/reference/supervisor/bandwidth-reduction.md) for a decent rendering.

rosswesleyporter commented:
First, this is great! Very useful. Very clear.

Suggest emphasizing the Supervisor version and the balenaOS version. From reading, it's not immediately obvious that there is a substantial benefit to using 2.107.10 & 14.4.4 or later. But if I understand correctly there is a big benefit. Could add a sentence above the table. Or perhaps put the versions in the table itself.

Above the table you may want to mention that the settings are reversible. For instance, the VPN can be turned back on. For some customers, it may sound scary to turn off the VPN.

This may not be important enough to include, but turning off Cloudlink turns off the ability to HUP.

In Device Logging, perhaps change "or balena system" to "or the balena system"

Again, this is great!

kb2ma commented:
Glad to hear the doc is clear and useful, @rosswesleyporter . I can emphasize cautions and required versions. Speaking of versions, I'm going to move this PR back to draft until there is a more recent release for RPi4. I manually added a fix that's in a later version of balenaOS than the version currently in the doc, 2.107.10.

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
-| `BALENA_SUPERVISOR_POLL_INTERVAL`      | 600000 to 86400000 | API Poll interval in milliseconds                  | 900000  | v1.3.0                            |
-| `BALENA_SUPERVISOR_LOG_CONTROL`        | true/false         | Enable / Disable logs from being sent to API       | true    | v1.3.0                            |
-| `BALENA_SUPERVISOR_HARDWARE_METRICS`   | true/false         | Enable / Disable device hardware metrics reporting | true    | v12.8.0                           |
+You may control these features from the fleet/device configuration tabs on the dashboard and as device variables. See the dashboard [docs][device-configuration] and the Configuration List [page][configuration-list]. Below the table we describe our measurement tool so you can repeat the tests.  Finally, be sure to read the [Side Effects](#side-effects--warnings) section for the impact of turning o ...
```
```suggestion
You may control these features from the fleet/device configuration tabs on the dashboard and as device variables. See the dashboard [docs][device-configuration] and the Configuration List [page][configuration-list]. Below the table we describe our measurement tool so you can repeat the tests.  Finally, be sure to read the [Side Effects](#side-effects--warnings) section for the impact of turning off each feature, particularly Cloudlink.
```

kb2ma replied:
Thanks @dfunckt , I have removed "VPN" from references to Cloudlink.

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
--------------------
+| Feature / Variable Name | Monthly Bandwidth if enabled | Measurement Notes |
+| ------- | --------- | -------- |
+| Cloudlink VPN<br>`BALENA_SUPERVISOR_VPN_CONTROL` | 80 MB | No use of VPN, just bandwidth required to maintain the connection  |
```
```suggestion
| Cloudlink<br>`BALENA_SUPERVISOR_VPN_CONTROL` | 80 MB | No use of VPN, just bandwidth required to maintain the connection  |
```

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
 
-Example minimum bandwidth settings
-----------------------------------
+**Cloudlink VPN** Provides the ability to send instantaneous updates to the device. Turning off cloudlink has several downsides: (1) configuration and variable updates are delayed by the device API polling interval; (2) CLI access is disabled via the web terminal or SSH; (3) device's public URL is not accessible. To be clear, you can re-enable Cloudlink via API or console after turning it off. Ho ...
```
```suggestion
**Cloudlink** Provides the ability to send instantaneous updates to the device. Turning off cloudlink has several downsides: (1) configuration and variable updates are delayed by the device API polling interval; (2) CLI access is disabled via the web terminal or SSH; (3) device's public URL is not accessible. To be clear, you can re-enable Cloudlink via API or console after turning it off. However you must wait for the next API poll from the device before the link is re-established.
```

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
 
-Example minimum bandwidth settings
-----------------------------------
+**Cloudlink** Provides the ability to send instantaneous updates to the device. Turning off cloudlink has several downsides: (1) configuration and variable updates are delayed by the device API polling interval; (2) CLI access is disabled via the web terminal or SSH; (3) device's public URL is not accessible. To be clear, you can re-enable Cloudlink via API or console after turning it off. Howeve ...
```
```suggestion
**Cloudlink** allows the device to be instantly notified of state changes. Turning off Cloudlink has several consequences:

- CLI access via the web terminal or SSH is disabled.
- Updates to configuration, environment variables and other state are not picked up by the device immediately, but are delayed instead until the next time the device polls the API.
- The Device Public URL feature is not available.

You can re-enable Cloudlink via API or console after turning it off. The device will re-establish the link the next time it polls the API.
```

------

Also, what do you mean with "via API or console"? Can you please elaborate as I don't think it's clear to the reader.

kb2ma replied:
Just changed that to "dashboard", which works better for Cloudlink.

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
+**Cloudlink** Provides the ability to send instantaneous updates to the device. Turning off cloudlink has several downsides: (1) configuration and variable updates are delayed by the device API polling interval; (2) CLI access is disabled via the web terminal or SSH; (3) device's public URL is not accessible. To be clear, you can re-enable Cloudlink via API or console after turning it off. Howeve ...
 
-The following settings lead to data usage of approximately 1.3MB per month:
+**Device Metrics** [Device metrics][device-metrics] such as CPU temperature and memory usage won't be reported to the API if turned off, and so won't be displayed on the summary page of the device dashboard.
```
```suggestion
[**Device Metrics**][device-metrics] controls whether metrics such as CPU temperature and memory usage, are reported to the API and made available for inspection on the summary page of the device dashboard. If turned off, apart from the network bandwidth savings, the respective section on the dashboard is hidden.
```

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
-* Change `BALENA_SUPERVISOR_POLL_INTERVAL` to 24 hours (86400000 ms)
-* Disable `BALENA_SUPERVISOR_LOG_CONTROL`
-* Disable `BALENA_SUPERVISOR_HARDWARE_METRICS`
+**API Poll Interval** Defines the time interval between device checks for updates to the fleet, for example new code pushes, environment variables changes, or cloudlink control changes for the device. If Cloudlink is disabled, lengthening this interval increases the latency of propagating these updates to the device.
```
```suggestion
**API Poll Interval** defines the time interval between device checks for updates to the fleet, for example new code pushes, environment variables changes, or Cloudlink control changes for the device. If Cloudlink is disabled, lengthening this interval increases the latency of propagating these updates to the device.
```

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
-* Disable `BALENA_SUPERVISOR_HARDWARE_METRICS`
+**API Poll Interval** Defines the time interval between device checks for updates to the fleet, for example new code pushes, environment variables changes, or cloudlink control changes for the device. If Cloudlink is disabled, lengthening this interval increases the latency of propagating these updates to the device.
 
+**NetworkManager connectivity check** When enabled for devices with multiple network interfaces, NetworkManager uses routing metrics to avoid use of an interface that fails the check. This adaptability is crucial to retain connectivity. However for a single network interface, NetworkManager can provide only interface status from the check via DBus.
```
```suggestion
**NetworkManager connectivity check**, when enabled for devices with multiple network interfaces, instructs NetworkManager to use routing metrics to avoid using an interface that fails the check. This adaptability is crucial to retain connectivity. However for a single network interface, NetworkManager can provide only interface status from the check via DBus.
```

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
 
+**NetworkManager connectivity check** When enabled for devices with multiple network interfaces, NetworkManager uses routing metrics to avoid use of an interface that fails the check. This adaptability is crucial to retain connectivity. However for a single network interface, NetworkManager can provide only interface status from the check via DBus.
+
+**Device Logging** Any logs written by the user container or the balena system are not sent to the dashboard when this variable is set to false.
```
```suggestion
**Device Logging** controls whether logs written by your application or components of balenaOS are sent to the dashboard or not.
```

dfunckt commented on pages/reference/supervisor/bandwidth-reduction.md:
```diff
@@ -4,52 +4,47 @@ title: Reduce bandwidth usage
+
+**Device Logging** Any logs written by the user container or the balena system are not sent to the dashboard when this variable is set to false.
+
+**Cloudlink connectivity check** Defines the device's ability to test and indicate (via an LED when available) that it has issues with connectivity.
```
```suggestion
**Cloudlink connectivity check** defines the device's ability to test and indicate (via an LED when available) that it has issues with connectivity.
```

kb2ma commented:
Thanks for the improvements to the Side Effects section, @dfunckt . I agree that many of them read better and incorporated in the latest commit. 

I did continue to start the explanatory text as a new sentence. I also added a colon ":" after the bold heading. I wanted the words in bold to refer to the feature back up to the table. I think trying to make a sentence begin with the heading adds too much burden on those words.

kb2ma commented:
When merge this PR, must follow up with user for #2126. Also must review patterns related to bandwidth.

kb2ma commented:
@vipulgupta2048 , this PR is ready for your detailed review.

vipulgupta2048 reviewed (commented):
Looking fine-o-fine. I just pushed a new commit containing some of the changes I felt that could be helpful. The only review I make is to reduce commits down to 1 or 2 ideally. Since most of the changes are about improving the content of the same only. 
Feel free to self-certify after that

kb2ma commented:
@balena-ci I self-certify!
